### PR TITLE
Add scoped NVTX ranges for improved profiling

### PR DESCRIPTION
**Pull request by esseivaju**
Use the NVTX library instead of cuda profiler. For now this implements the bare minimum functionalities to be useful. We'd probably want to use [registered strings](https://nvidia.github.io/NVTX/doxygen/group___s_t_r_i_n_g___r_e_g_i_s_t_r_a_t_i_o_n.html) and possibly allow to specify a payload / category?

### CODE REVIEW

File: src/corecel/sys/ScopedDeviceProfiling.hh [hunk 1]
```diff
@@ -30,20 +32,14 @@ class ScopedDeviceProfiling
 {
   public:
     // Activate profiling
-    ScopedDeviceProfiling();
+    ScopedDeviceProfiling(std::string_view name);
```
Comment (sethrj) [inline src/corecel/sys/ScopedDeviceProfiling.hh]:
```suggestion
    explicit ScopedDeviceProfiling(std::string_view name);
```

Is `name` supposed to be the subdomain name (like `celeritas::app`)? Or is it a segment of code?

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc [hunk 2]
```diff
@@ -20,24 +20,38 @@
 
 namespace celeritas
 {
+namespace
+{
+nvtxDomainHandle_t domain_handle()
+{
+    static nvtxDomainHandle_t domain = nvtxDomainCreateA("Celeritas");
```
Comment (sethrj) [inline src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: Should we have the domain match the namespace casing (`celeritas`?) Or is it standard practice to have the domain fancy-cased?

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc [hunk 3]
```diff
@@ -47,16 +61,9 @@ ScopedDeviceProfiling::ScopedDeviceProfiling()
  */
 ScopedDeviceProfiling::~ScopedDeviceProfiling()
 {
-    if (activated_)
+    if (celeritas::device())
```
Comment (sethrj) [inline src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: Since it's possible to clear the device between construction and destruction, it might still be good to keep the `activated_` bool for safety?

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc [hunk 4]
```diff
@@ -20,24 +20,38 @@
...
-        {
-            CELER_CUDA_CALL(cudaProfilerStart());
-        }
-        catch (RuntimeError const& e)
-        {
-            CELER_LOG(error) << "Failed to start profiling: " << e.what();
-            return;
-        }
-        activated_ = true;
+        nvtxEventAttributes_t attributes_ = attributes(name.data());
```
Comment (sethrj) [inline src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: There's no guarantee that `name` is going to have a `\0` after it, and the attributes take a `const char` rather than a pointer+size. So this unfortunately has to use `std::string const&` (or const char* explicitly) to be safe, and `c_str()` here.

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc [hunk 5]
```diff
@@ -20,24 +20,38 @@
...
 namespace celeritas
 {
+namespace
+{
+nvtxDomainHandle_t domain_handle()
+{
+    static nvtxDomainHandle_t domain = nvtxDomainCreateA("Celeritas");
+    return domain;
+}
+nvtxEventAttributes_t attributes(char const* name)
```
Comment (sethrj) [inline src/corecel/sys/ScopedDeviceProfiling.cuda.cc]:
```suggestion
nvtxEventAttributes_t make_attributes(char const* name)
```
since it's returning a new object by value as opposed to accessing something

File: src/corecel/sys/ScopedDeviceProfiling.hh (hunk 1 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedDeviceProfiling.hh]: `name` refers to a segment of code. The subdomain would be the `category` where you could group range (e.g. I/O, malloc, ...)

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc (hunk 2 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: We can match the namespace

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc (hunk 3 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: NVTX can be used to profile host functions as well so maybe we want to have this class more generic and not only limited to profiling Kernels?

Comment (sethrj) [reply src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: If it's safe to call outside of creating a CUDA context then yeah we can eliminate the `device` checks. If we have to have a device, I think saving a flag would be safer.

Comment (esseivaju) [reply src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: It is safe to call without a CUDA context, in fact, if we only include `nvToolsExt.h` there's no dependency on cuda.

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc [hunk 6]
```diff
@@ -22,20 +27,75 @@ namespace celeritas
...
+    {
+        auto handle
+            = nvtxDomainRegisterStringA(domain_handle(), message.c_str());
+        message_registry().insert({message, handle});
+        return handle;
+    }
+    else
+    {
+        return message_handle->second;
+    }
```
Comment (sethrj) [inline src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: The solution is to just `insert` with a `nullptr` for the handle, and if insertion succededed to register it and save in the returned iterator.
```suggestion
    auto [iter, inserted] = message_registry().insert({message, nullptr});
    if (inserted)
    {
        iter->second
            = nvtxDomainRegisterStringA(domain_handle(), message.c_str());
    }
    CELER_ENSURE(iter->second);
    return iter->second;
```

File: src/corecel/CMakeLists.txt [hunk 7]
```diff
@@ -58,8 +58,9 @@ if(CELERITAS_USE_CUDA OR CELERITAS_USE_HIP)
   )
 endif()
 if(CELERITAS_USE_CUDA)
+  # we currently depend on CUDA install including the nvtx header
   list(APPEND SOURCES
-    sys/ScopedDeviceProfiling.cuda.cc
+    sys/ScopedProfiling.nvtx.cc
```
Comment (sethrj) [inline src/corecel/CMakeLists.txt]: Unless we add a new `find_package(NVTX)` and `CELERITAS_USE_NVTX`  let's leave this as `.cuda`

File: src/corecel/sys/ScopedProfiling.hh [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+class ScopedProfiling
+{
+  public:
+    using Input = ScopedProfilingInput;
+    // Activate profiling
+    explicit ScopedProfiling(Input input);
+    // Deactivate profiling
+    ~ScopedProfiling();
+    // RAII semantics
+    void* operator new(std::size_t count) = delete;
```
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh]: I've never seen this pattern before... is it necessary? Do we need to add it to all our RAII classes? For comparison, [scoped_lock](https://en.cppreference.com/w/cpp/thread/scoped_lock) deletes the assignment operator but doesn't delete `new`.

File: src/corecel/sys/ScopedProfiling.hh [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ * This is useful for wrapping specific code fragment in a range for profiling,
+ * e.g. ignoring of VecGeom instantiation kernels, profiling a specific action
+ * or loop on the CPU.
+ * TODO: Template ScopedProfiling over profiling backend if we need to add a
+ * new one
+ */
+class ScopedProfiling
+{
+  public:
+    using Input = ScopedProfilingInput;
```
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh:50]: Keep the style consistent...
```suggestion
    //!@{
    //! \name Type aliases
    using Input = ScopedProfilingInput;
    //!@}

  public:
```

File: src/corecel/sys/ScopedProfiling.nvtx.cc [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+    {
+        std::shared_lock lock(mutex);
+        auto message_handle = message_registry().find(message);
+        if (message_handle != message_registry().end())
+        {
+            return message_handle->second;
+        }
+    }
+    // we did not find the handle, insert it
```
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.nvtx.cc]: Try to use leading case for comments
```suggestion
    // We did not find the handle: try to insert it
```

File: src/corecel/sys/ScopedProfiling.hh [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    // Activate profiling
+    explicit ScopedProfiling(Input input);
```
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh:57]: Should we always have to specify all the input parameters, or are they just options? I think some of the code would be cleaner if we also provide a "name-only" constructor for the simple use cases.

File: src/corecel/sys/ScopedDeviceProfiling.cuda.cc (hunk 6 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: @sethrj Actually, we can't really `CELER_ENSURE` that iter->second is non-null. If we're not running the application through a tool that can handle nvtx, i.e. ncu with --nvtx opthon, then all API calls will return 0 ([ref](https://github.com/NVIDIA/NVTX#how-much-overhead-does-nvtx-add-to-my-code)).

Also we can't really make any assumption on what values are returned. I tried having multiple nested ranges and printing each value returned by `nvtxDomainRegisterStringA`. The first call returns 0, the second call returns 1 so they're not really pointers as you'd expect ([relevant doc](https://github.com/NVIDIA/NVTX#give-dont-take)).

TLDR; just don't call CELER_ENSURE and trust the implementation?

Comment (sethrj) [reply src/corecel/sys/ScopedDeviceProfiling.cuda.cc]: Oh interesting. I guess "handle" doesn't mean "pointer" in this case. (Maybe replace `nullptr` with `{}`?) That's really good to know. Could you add a note in the implementation that the API calls may fail if the profiler is disabled?

File: src/corecel/sys/ScopedProfiling.hh (hunk 8 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.hh]: It's not technically needed but per the name of the class, dynamically allocating a new instance is not "scoped profiling", so does it make sense to allow it? The C++ API of nvtx does that for their implementation of the scoped_range.

File: src/corecel/sys/ScopedProfiling.hh (hunk 11 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.hh:57]: A constructor with just the name definitely makes sense as most of the time we'd want to only specify that.

File: src/corecel/sys/ScopedProfiling.hh (hunk 8 above)
Comment (sethrj) [reply src/corecel/sys/ScopedProfiling.hh]: I think `delete` and such should be to prevent *accidental* misuses, since it's not possible for us to prevent all *intentional* misuse.

Besides, if wanted to (for example) put it in a `unique_ptr` or `shared_ptr` to change/share the scope at runtime, I think that's a reasonable use case. A dumb user could `new` it and never release it, and then it just becomes an open range? Which seems like the expected behavior.

Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.hh]: In general I agree, but for this class and the stack semantics of nested ranges I'm not sure it would be legal to outlive the scope. Suppose we have `Scope1` active from a caller function and we instantiate `Scope2` and have it managed by a `unique_ptr` to outlive `Scope1`.  Then, the `Scope1` dtor would, I assume, end the range from the nested scope. 

Also those range are thread_local so passing the ownership to another thread shouldn't be done which could easily happen with smart pointers or heap allocation.

File: src/corecel/sys/ScopedProfiling.hh [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+{
+
+//---------------------------------------------------------------------------//
+/*!
+ * Input arguments for the nvtx implementation.
+ */
+struct ScopedProfilingInput
+{
+    std::string name;  //!< Name of the range
+    uint32_t color{0xFF00FF00};  //!< ARGB
```
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh]: Requires `#include <cstdint>` and since it's unsigned:
```suggestion
    uint32_t color{0xFF00FF00u};  //!< ARGB
```

File: src/corecel/sys/ScopedProfiling.hh [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    explicit ScopedProfiling(std::string const& name);
+    // Deactivate profiling
+    ~ScopedProfiling();
+    // Can't outlive the scope it was created in to correctly match each push /
+    // pop. Must strictly follows RAII semantics (automatic storage duration)
+    void* operator new(std::size_t) = delete;
+    ScopedProfiling(ScopedProfiling const&) = delete;
+    ScopedProfiling& operator=(ScopedProfiling const&) = delete;
+    ScopedProfiling(ScopedProfiling&&) = delete;
+    ScopedProfiling& operator=(ScopedProfiling&&) = delete;
```
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh]: Remove the `operator new` specialization, and the `&&` instances are automatically deleted if `&` is deleted so you can omit those.

**Review by sethrj — APPROVED** [review]
Nice work Julien. Do you want to add these to the individual actions/kernel launches in a separate PR?

File: src/corecel/sys/ScopedProfiling.hh (hunk 13 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.hh]: I can see allowing `operator new` even though using it is asking for trouble. 

Regarding the move operations, should we be explicit and follow [C.21](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all)?

**Comment by esseivaju** [comment]
Yeah, I can do that in a separate PR though there is one thing I don't like; If nvtx isn't enabled, you still end up paying a small cost calling `message_handle_for`  and locking mutexes so sprinkling the codebase with `ScopedDevice` has a cost...

File: src/corecel/sys/ScopedProfiling.hh (hunk 13 above)
Comment (sethrj) [reply src/corecel/sys/ScopedProfiling.hh]: Good call. It looks like `ScopedMem`, `Stream`, and `ScopedSignalHandler` all need that as well, and a few "scoped" classes are entirely missing the deleted copy/assign:
- src/celeritas/ext/ScopedGeantExceptionHandler.hh
- src/celeritas/ext/ScopedGeantLogger.hh
- src/celeritas/ext/ScopedRootErrorHandler.hh
- src/corecel/io/ScopedStreamFormat.hh
- src/corecel/io/ScopedStreamRedirect.hh
- src/corecel/io/ScopedTimeAndRedirect.hh
- src/corecel/io/ScopedTimeLog.hh
- src/corecel/sys/ScopedMpiInit.hh

Should those updates be a separate PR too? If you haven't pushed the restoration of `..&&) = delete`  I think we should do them all in a separate best-practices merge.

**Comment by sethrj** [comment]
Maybe to avoid the cost, we use the same trick as `ScopedSignalHandler` and use a global environment variable to disable it?

File: src/corecel/sys/ScopedProfiling.hh (hunk 13 above)
Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.hh]: sure, we can do that in a separate PR.

**Comment by esseivaju** [comment]
yeah, that's sounds like a good solution. I'd say in this case we want to explicitly enable profiling with `CELER_ENABLE_PROFILING`

**Review by sethrj — APPROVED** [review]
